Added the Request controller

MilesDev91 opened the pull request:
### What this PR does:
- Adds a controller to the server for Requests

### Any information needed to test this feature:
- View the request.js file in /server/controllers and /server/routes
- You can test the API with postman by:
- - Logging in with the /auth/login by sending the email and password in a body object (you may have to register a new user first, which you can do in the application or through postman at /auth/register)
- - Once you are authorized, you may test the /requests/create, /requests/edit/:requestId, and /requests/?userType= routes. 
- - userType must be either "sitter" or "requester". Check the request model in /server/models to see the proper data to send.

### Any issues with the current functionality:
- There is no integration with the front end.

closes #8

### Review thoughts:
- I haven't worked with controllers since doing MVC in ASP.net about three years ago, so this is very new to me. I gathered what I could from various documentation and the other controllers.
- Is the extra update method useful? I understand we can edit the status with the PUT method, but we edit status specifically more often than anything else NOTE: removed the PATCH method.

MilesDev91 committed:
Create request.js

Created the request controller.

Added methods to create, edit, and get requests. Also added a method to specifically update the request status.

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,79 @@
+// @desc Create new request
+// @access Public
+exports.createRequest = asyncHandler(async (req, res, next) => {
+  const request = await Request.create(req.body);
+
+  if (request) {
+    res.status(200).json({
+      success: {
+        request: {
+          ...request
+        }
+      }
+    })
+  } else {
+    res.status(400);
+    throw new Error("Invalid request data")
+  }
```
[Model.create](https://mongoosejs.com/docs/api.html#model_Model.create) returns a promise that either resolves into the document(s) being created, or rejects it with the encountered error(s). And as stated by mdn [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#description): 
> If the Promise is rejected, the await expression throws the rejected value

So the `else` block here will never execute. We can either:
- Remove the entire `else` block. This is okay since we have [`asyncHandler`](https://www.npmjs.com/package/express-async-handler/v/1.0.2) catching any thrown errors on our behalf, and the [error midleware](https://github.com/hatchways/team-pine/blob/request_controller/server/middleware/error.js) wired in [`app.js`](https://github.com/hatchways/team-pine/blob/c43e5dd58441cba5201370c3afc496a85fcf6c25/server/app.js#L62) handling thrown errors.
- Or we can wrap it in a `try..catch` block if we want to handle the error ourselves (this is useful if we want to send a status code other than `500`).

MilesDev91 replied:
Ok I will just remove it in that case. If we need another status code we can add it later.

MilesDev91 replied:
Removed!

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,79 @@
+// @desc edit request
+// @access Public
+exports.editRequest = asyncHandler(async (req, res, next) => {
+  const request = await Request.findOne({ _id: req.request._id })
```
I think there is a bit of a problem with how we expect to use this endpoint:

It's common to use a [route param](https://expressjs.com/en/guide/routing.html#route-parameters) to identify a single resource (e.g. the request being updated). This would look like `req.params.requestId` in your controller. We can then reserve `req.body` to contain either all the remaining fields in the document or only the ones being updated. 

There is also a potential security concern here but I will detail it after we implement the routes for the controllers.

Lastly, while `findOne` works just as well, it would have been more accurate to use [`model.findById`](https://mongoosejs.com/docs/api.html#model_Model.findById) here instead 👍🏼

MilesDev91 replied:
Ok!

MilesDev91 replied:
Ok I implemented the suggestions.

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,79 @@
+exports.editRequest = asyncHandler(async (req, res, next) => {
+  const request = await Request.findOne({ _id: req.request._id })
+  
+  if (!request) {
+    res.status(404);
+    throw new Error("Request doesn't exist");
+  }
```
👍🏼

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,79 @@
+    res.status(404);
+    throw new Error("Request doesn't exist");
+  }
+  request.set(req.body);
```
This should be a safe operation due to the [`strict` option](https://mongoosejs.com/docs/guide.html#strict), but I'd still recommend testing this endpoint with payloads that include fields that are not in the `Request` schema 👍🏼

MilesDev91 replied:
Would passing the req.body values into a new Request object then setting that be a good solution?

MilesDev91 replied:
Ok! Tested it and it didn't add anything extra when I gave it extra bogus values, but edited properly otherwise.

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,79 @@
+// @desc get requests of logged in user
+// @access Private
+exports.getUserRequests = asyncHandler(async (req, res, next) => {
+  const requests = await Request.where('requester', req.userId);
```
I don't think it's enough to return only the requests where the user is the owner/requester. What if we want to get the requests where the authenticated user is the sitter? That would be usefult for "manage bookings" page for the sitters for example. I think we can either return both, or better, implement [query params](http://expressjs.com/en/api.html#req.query) to let the client pick.

Also, there is a small typo below:
```suggestion
  const requests = await Request.where('requester', req.user.id);
```

MilesDev91 replied:
Added a query param for user type.

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,79 @@
+exports.getUserRequests = asyncHandler(async (req, res, next) => {
+  const requests = await Request.where('requester', req.userId);
+
+  if (!requests) {
+    res.status(404);
+    throw new Error("User has no requests or is not logged in")
+  }
```
There are a couple of issues with this:
- [`model.where/find`](https://mongoosejs.com/docs/api/model.html#model_Model.where) will return an array when it suceeds (sometimes empty, sometimes not). And if it fails (i.e. the query rejects), the `await` keyword will throw an error as mentioned before. So in no case will `!requests` evaluate to true (remember that empty arrays are [truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) too)!
- If the list of requests for the user is empty, we shouldn't respond with a `404`. `404` implies we couldn't locate the resource, which is not true in this case - we located it all fine but it's just empty. So it would be more appropriate to just send the empty collection with a success status code 👍🏼 (you can find more on this [here](https://stackoverflow.com/a/13367198/15220729))

MilesDev91 replied:
Makes sense

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,79 @@
+// @desc edit request status
+// @access Public
+exports.acceptRequest = asyncHandler((req, res, next) => {
+  const request = await Request.findOne({_id: req.request._id})
```
Similar comments to:
> I think there is a bit of a problem with how we expect to use this endpoint:
>
> It's common to use a [route param](https://expressjs.com/en/guide/routing.html#route-parameters) to identify a single resource (e.g. the request being updated). This would look like `req.params.requestId` in your controller. We can then reserve `req.body` to contain either all the remaining fields in the document or only the ones being updated. 
>
> There is also a potential security concern here but I will detail it after we implement the routes for the controllers.
> 
> Lastly, while `findOne` works just as well, it would have been more accurate to use [`model.findById`](https://mongoosejs.com/docs/api.html#model_Model.findById) here instead 👍🏼

ugur-akin commented on server/controllers/request.js:
```diff
@@ -0,0 +1,79 @@
+exports.acceptRequest = asyncHandler((req, res, next) => {
+  const request = await Request.findOne({_id: req.request._id})
+
+  if (!request) {
+    res.status(404);
+    throw new Error("No request found")
+  }
```
👍🏼

ugur-akin reviewed (changes requested):
Hey Miles, good work on the controllers! I left some feedback below, have a look and feel free to ask for further clarifications or share your thoughts 👍🏼 

One thing I noticed was that we didn't define the routes for the controllers, there are some examples of this under the [`routes`](https://github.com/hatchways/team-pine/tree/request_controller/server/routes) directory. Once we declare them, we should also use the router in [`app.js`](https://github.com/hatchways/team-pine/blob/c43e5dd58441cba5201370c3afc496a85fcf6c25/server/app.js#L45-L47) 👍🏼 

I also noticed the model file for the request was included as a change in this PR, but this isn't ideal because we are tracking its changes in a different issue/PR. To keep things simple, I would recommend [editing the base](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request) of the PR so the diff only shows the changes relevant to #8 

Thanks for all your work and for communicating your PR in a well-formatted manner 🚀

ugur-akin commented:
> Is the extra update method useful? I understand we can edit the status with the PUT method, but we edit status specifically more often than anything else.

If we had a single `PATCH` method that updated all the fields given in the payload, it could handle both cases. 

Apart from that, I would warn against optimizing things prematurely - an entire `PUT` request might just be fine for our purposes (I strongly recommend reading more about this [here](https://softwareengineering.stackexchange.com/a/80092)). The most apparent downside of having more endpoints would be confusion, but if we document things well that shouldn't be a problem either. So really, both approaches are fine 😅

MilesDev91 committed:
Modified Request controller and routes

Added routes with protection for request controller. Implemented the routes in app.js

Added validation for user type query params.

Removed the patch route in the request controller. It just isn't necessary yet and will likely hurt more than help.

Cleaned up some unnecessary error codes and modified some queries in the controller.

Fixed False typo in the request model that was missed in the request model pr.

MilesDev91 commented:
I understand how the controllers and routes work now. Thanks for the suggestions. Tested the endpoints with postman and they work as expected. Let me know if anything looks out of place!

MilesDev91 committed:
Update request.js

Clarified the query param in the route comment for getUserRequests

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,51 @@
+// @desc Create new request
+// @access Public
+exports.createRequest = asyncHandler(async (req, res, next) => {
+  const request = await Request.create(req.body);
```
It is a good practice to destructure the request body to only allow selected fields for security purposes (for example, the user can accept the request manually before it even gets to the sitter)

```
const { sitterId, startDate, endDate } = req.body
```

MilesDev91 replied:
Ah I see

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,51 @@
+    res.status(404);
+    throw new Error("Request doesn't exist");
+  }
+  request.set(req.body);
```
We would want only the sitter to edit a request or restrict the owner from editing fields like paid, accepted et cetera :)

rajivtitus commented on server/routes/request.js:
```diff
@@ -0,0 +1,13 @@
+const protect = require("../middleware/auth");
+const { createRequest, editRequest, getUserRequests } = require("../controllers/request");
+
+router.route("/create").post(protect, createRequest);
```
```suggestion
router.route("/").post(protect, createRequest);
```

Nice routes. This route can further be simplified though as the HTTP verb indicates the purpose of that route 👍🏽

MilesDev91 replied:
Ok!

rajivtitus commented on server/validate.js:
```diff
@@ -30,3 +30,15 @@ exports.validateLogin = [
   }
 ];
+
+exports.validateUserTypeQuery = [
```
👍🏽

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,51 @@
+// @desc get requests of logged in user
+// @access Private
+exports.getUserRequests = asyncHandler(async (req, res, next) => {
+  const requests = await Request.where(req.query.userType, req.body.user.id);
```
Let's refrain from using the req.body in a GET request. A payload within a GET request message has no defined semantics; sending a payload body on a GET request might cause some existing implementations to reject the request.

MilesDev91 replied:
Makes sense

rajivtitus reviewed (commented):
Impressive work despite no exposure to express.js 👍🏽 
Some improvements can be made, I've made comments where necessary.

MilesDev91 committed:
Updated request controller

Simplified one route and added a param to another.

Updated controllers to avoid directly setting to the req.body for security reasons. Enforced the sitter user type for editing requests as well.

Changed the get controller to avoid using a request body.

MilesDev91 committed:
Update controller and route

We pass in user id through req.user.id, so no parameter needed.

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,64 @@
+// @access Public
+exports.editRequest = asyncHandler(async (req, res, next) => {
+  // Until we implement cancelling bookings, owners shouldn't be able to edit bookings
+  if (req.query.userType != "sitter") {
```
Any user will be able to send these optional query parameters to edit this request :)
Please look at the example below to further understand how we would go about this issue:
```
const userId = req.user.id
const requestId = req.params.requestId

if (!mongoose.Types.ObjectId.isValid(requestId)) {
   return res.status(400).send("Bad Request");
 }

const request = Request.findById(requestId)

if (request.sitter === userId) {
   //This implies that the logged in user is a sitter and the service provider in this request.
   He can edit (accept/decline) this request
} else {
   res.status(403)
   throw new Error ("You are not authorized to perform this operation")
}

```

MilesDev91 replied:
Ya I didn't understand how user object was getting passed in without params until Gurkirt told me last night, so I changed it back. Matching the sitter field to the user Id seems pretty obvious in retrospect. Woops. I will do similar changes to the GET as well.

rajivtitus commented on server/controllers/request.js:
```diff
@@ -0,0 +1,64 @@
+    throw new Error("Request doesn't exist");
+  }
+
+  const { status, paid } = req.body
```
```suggestion
  const { status } = req.body
```

Let's refrain from changing the `paid` status within this controller. We can do it through the controller in which we process payments 😄

MilesDev91 replied:
Oh boy I have no clue why this didn't cross my mind as bad.

rajivtitus reviewed (commented):
Good work Miles! Just left a comment about a change in the implementation, please let me know if you have any additional questions. Once the comment has been addressed, I think this should be good to go 👍🏽

MilesDev91 committed:
Updated request controller and route

Removed passing things like user type and user id through params, and adjusted the code/routes to maintain functionality.

Removed setting paid status through request controller.

rajivtitus commented on server/validate.js:
```diff
@@ -30,3 +30,15 @@ exports.validateLogin = [
   }
 ];
+
+exports.validateUserTypeQuery = [
```
I don't think we'll need this anymore though but you can leave it in for now 😄

MilesDev91 replied:
Ya I will remove it and replace it with other validation.

rajivtitus reviewed (approved):
LGTM, nice work 👍🏽 

As a bonus, you can look into validating the request body and route params. Here's a nice article on how to do so using express-validator: 
https://www.freecodecamp.org/news/how-to-make-input-validation-simple-and-clean-in-your-express-js-app-ea9b5ff5a8a7/

MilesDev91 committed:
Added a bit of validation for request creation

MilesDev91 committed:
Merge branch 'request_model' into request_controller
